File: ios/Sources/Omnesis/UI/AccessAuthorizationComponents.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation

#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI

@available(iOS 17.0, *)
struct AccessAuthorizationCodeEntry: View {
    @Binding var code: String
    let loading: Bool
    let errorMessage: String?
    let onLookup: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Theme.Spacing.lg) {
            VStack(alignment: .leading, spacing: Theme.Spacing.sm) {
                Image(systemName: "person.badge.key")
                    .font(.system(size: 34))
                    .foregroundStyle(Theme.accent)
                Text("Enter the code shown by the app that wants to connect.")
                    .font(.system(size: 20, weight: .semibold))
                    .foregroundStyle(Theme.textPrimary)
                Text(
                    "The notification contains no request details. " +
                        "Omnesis shows who is connecting only after your phone verifies this code."
                )
                .font(.system(size: 14))
                .foregroundStyle(Theme.textSecondary)
            }
            TextField("ABCD-EFGH", text: $code)
                .textInputAutocapitalization(.characters)
                .autocorrectionDisabled()
                .font(.system(.title3, design: .monospaced, weight: .semibold))
                .padding(Theme.Spacing.md)
                .background(Theme.bgSecondary)
                .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.medium))
                .accessibilityLabel("Authorization code")
            if let errorMessage {
                Text(errorMessage)
                    .font(.system(size: 13, weight: .medium))
                    .foregroundStyle(Theme.warning)
                    .accessibilityLabel("Authorization lookup failed")
            }
            Button(action: onLookup) {
                HStack {
                    if loading { ProgressView().tint(.white) }
                    Text(loading ? "Checking…" : "Continue")
                        .font(.system(size: 15, weight: .semibold))
                }
                .frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)
            .tint(Theme.accent)
            .disabled(loading || code.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
            Spacer()
        }
        .padding(Theme.Spacing.lg)
    }
}

@available(iOS 17.0, *)
struct AccessAuthorizationOutcomeContent: View {
    let outcome: AccessAuthorizationOutcome
    let clientName: String?
    let onDone: () -> Void

    var body: some View {
        let presentation = accessAuthorizationOutcomePresentation(
            outcome: outcome,
            clientName: clientName
        )
        VStack(spacing: Theme.Spacing.lg) {
            Image(systemName: presentation.symbol)
                .font(.system(size: 56))
                .foregroundStyle(presentation.approved ? Theme.success : Theme.warning)
            Text(presentation.title)
                .font(.title2.weight(.semibold))
            Text(presentation.detail)
                .multilineTextAlignment(.center)
                .foregroundStyle(.secondary)
            Button("Done", action: onDone)
                .buttonStyle(.borderedProminent)
        }
        .padding(Theme.Spacing.xl)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

@available(iOS 17.0, *)
struct AccessAuthorizationStepIndicator: View {
    /// How much of each step's chip there is room to draw.
    private enum Density {
        /// Every step names itself in full.
        case full
        /// Every step still names itself, some of them abbreviated.
        case short
        /// Only the step being worked on names itself; the rest are numbers.
        case currentOnly
        /// Numbers alone.
        case numbers
    }

    let steps: [AccessAuthorizationStep]
    let current: AccessAuthorizationStep
    let compact: Bool
    /// Returning to a finished step.
    let onSelect: (AccessAuthorizationStep) -> Void

    /// The marker grows with the reader's type size: at accessibility sizes it
    /// is the only thing left of the chip, and it is the target you tap to go
    /// back, so it must not stay a 16pt dot beside 40pt text.
    @ScaledMetric(relativeTo: .caption2) private var markerSize: CGFloat = 16

    private var currentIndex: Int {
        steps.firstIndex(of: current) ?? 0
    }

    var body: some View {
        // At accessibility type sizes the titles cannot fit at any density, so
        // the numbers-only row is the only candidate offered.
        HStack(spacing: 0) {
            if compact {
                row(.numbers)
            } else {
                ViewThatFits(in: .horizontal) {
                    row(.full)
                    row(.short)
                    row(.currentOnly)
                    row(.numbers)
                }
            }
            Spacer(minLength: 0)
        }
    }

    private func row(_ density: Density) -> some View {
        HStack(spacing: 5) {
            ForEach(Array(steps.enumerated()), id: \.element) { index, item in
                chip(index: index, item: item, density: density)
            }
        }
    }

    @ViewBuilder
    private func chip(index: Int, item: AccessAuthorizationStep, density: Density) -> some View {
        let isCurrent = item == current
        let isComplete = index < currentIndex
        let showsTitle = density == .full
            || density == .short
            || (density == .currentOnly && isCurrent)
        let label = density == .full ? item.title : item.shortTitle
        Button {
            if isComplete { onSelect(item) }
        } label: {
            HStack(spacing: 5) {
                marker(index: index, isCurrent: isCurrent, isComplete: isComplete)
                if showsTitle {
                    Text(label)
                        .font(.caption2.weight(isCurrent ? .semibold : .regular))
                        .foregroundStyle(isCurrent ? Theme.textPrimary : Theme.textSecondary)
                        .lineLimit(1)
                        .fixedSize(horizontal: true, vertical: false)
                }
            }
            .padding(.horizontal, showsTitle ? 7 : 5)
            .padding(.vertical, 5)
            .background(isCurrent ? Theme.bgSecondary : Color.clear)
            .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.large))
            // The chip is small by design; the target you press is not.
            .frame(minHeight: Theme.tapTarget)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(!isComplete && !isCurrent)
        .accessibilityLabel(item.title)
        .accessibilityValue(isComplete ? "Completed" : isCurrent ? "Current step" : "Not started")
        .accessibilityAddTraits(isCurrent ? .isSelected : [])
    }

    /// The numbered disk: filled on the step being worked on, a checked ring
    /// once it is behind you, an outline while it is still ahead.
    private func marker(index: Int, isCurrent: Bool, isComplete: Bool) -> some View {
        ZStack {
            Circle()
                .fill(isCurrent ? Theme.accent : Color.clear)
            Circle()
                .strokeBorder(
                    isComplete ? Theme.success : isCurrent ? Color.clear : Theme.border,
                    lineWidth: 1
                )
            if isComplete {
                Image(systemName: "checkmark")
                    .font(.system(size: markerSize * 0.5, weight: .bold))
                    .foregroundStyle(Theme.success)
            } else {
                Text("\(index + 1)")
                    .font(.system(size: markerSize * 0.56, weight: .semibold))
                    .foregroundStyle(isCurrent ? Color.white : Theme.textMuted)
            }
        }
        .frame(width: markerSize, height: markerSize)
    }
}

/// One capability, drawn the same way wherever it is named.
///
/// A review that invents its own labels makes the owner check twice that the
/// thing they are about to approve is the thing they will be shown afterwards.
@available(iOS 17.0, *)
struct AccessCapabilityBadge: View {
    let capability: AccessCapability
    let tone: AccessCapabilityTone

    private var color: Color {
        switch tone {
        case .granted: Theme.success
        case .raw: Theme.danger
        case .unreviewed: Theme.warning
        case .notes: Theme.accent
        case .withheld: Theme.textMuted
        }
    }

    var body: some View {
        Text(capability.label.uppercased())
            .font(.caption2.weight(.semibold))
            .kerning(0.4)
            .foregroundStyle(color)
            .padding(.horizontal, 7)
            .padding(.vertical, 3)
            .overlay(
                Capsule().strokeBorder(
                    tone == .withheld ? Color.clear : color.opacity(0.45),
                    lineWidth: 1
                )
            )
            .opacity(tone == .withheld ? 0.35 : 1)
            .accessibilityLabel(accessibilityLabel)
    }

    private var accessibilityLabel: String {
        switch tone {
        case .withheld: "\(capability.label) not granted"
        case .unreviewed: "\(capability.label) granted, released without privacy review"
        default: "\(capability.label) granted"
        }
    }
}

/// Three fixed slots in a stable order, so a capability's position never moves
/// and "which of these has raw access" is answerable at a glance. A withheld
/// capability keeps its slot and goes faint rather than disappearing. The
/// badges sit in a row when it fits and stack when the text is too large for
/// one, so a label is never truncated or broken mid-word.
@available(iOS 17.0, *)
struct AccessCapabilityTriad: View {
    let holdings: AccessCapabilityHoldings

    var body: some View {
        ViewThatFits(in: .horizontal) {
            HStack(spacing: 5) {
                badges
                Spacer(minLength: 0)
            }
            VStack(alignment: .leading, spacing: 4) {
                badges
            }
        }
    }

    private var badges: some View {
        ForEach(AccessCapability.allCases, id: \.self) { capability in
            AccessCapabilityBadge(capability: capability, tone: holdings.tone(capability))
                .fixedSize(horizontal: true, vertical: false)
        }
    }
}

/// The one source list a scope runs under: which connected sources it may
/// read, and whether a source connected later joins them.
@available(iOS 17.0, *)
struct AccessSourceRuleEditor: View {
    let scope: AccessSourceScope
    @Binding var value: AccessSourceSelectionState
    let sources: [AccessSourceInstance]
    let store: AppStore
    @State private var query = ""

    private var availableSourceIds: Set<String> {
        Set(sources.filter(\.available).map(\.id))
    }

    private var allowedCount: Int {
        value.allowedSourceIds.intersection(availableSourceIds).count
    }

    /// No refusal while there is nothing to choose from: "select at least one
    /// source" in front of an empty list asks for something that is not there,
    /// and the empty state already says what to do instead.
    private var boundaryError: String? {
        guard !availableSourceIds.isEmpty else { return nil }
        return accessSourceBoundaryError(
            scope: scope,
            permitsAnyKnownSource: value.permitsAnyKnownSource(availableSourceIds),
            recordedSelectionCount: value.boundary(
                knownSourceIds: availableSourceIds
            ).sourceIds.count
        )
    }

    private var shownSources: [AccessSourceInstance] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !needle.isEmpty else { return sources }
        return sources.filter {
            $0.name.localizedCaseInsensitiveContains(needle)
                || $0.id.localizedCaseInsensitiveContains(needle)
        }
    }

    var body: some View {
        Section {
            if availableSourceIds.isEmpty {
                Text("No sources are connected. Connect a source before approving this access.")
                    .font(.footnote)
                    .foregroundStyle(.secondary)
            } else {
                HStack {
                    Text("Checked sources are allowed.")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                    Spacer(minLength: Theme.Spacing.sm)
                    Text("\(allowedCount) of \(availableSourceIds.count) allowed")
                        .font(.footnote.monospacedDigit())
                        .foregroundStyle(.secondary)
                }
                TextField("Search sources", text: $query)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                HStack {
                    Button("Allow all") { value.allowAll(availableSourceIds) }
                    Spacer()
                    Button("Block all") { value.blockAll(availableSourceIds) }
                }
                // Buttons in the same Form row otherwise inherit List's row-wide
                // tap behavior, which can invoke both actions from one tap.
                .buttonStyle(.borderless)
                // A checkbox rather than a switch: these are many rows of one
                // list being picked from, not a handful of independent
                // settings, and the whole row is the target. Android draws the
                // same row the same way.
                ForEach(shownSources) { source in
                    let allowed = value.allowedSourceIds.contains(source.id)
                    Button {
                        value.setAllowed(!allowed, sourceId: source.id)
                    } label: {
                        HStack(spacing: Theme.Spacing.sm) {
                            Image(systemName: allowed ? "checkmark.square.fill" : "square")
                                .font(.system(size: 16))
                                .foregroundStyle(allowed ? Theme.accent : Theme.textMuted)
                            SourceIconView(sourceId: source.id, store: store, size: 20)
                            VStack(alignment: .leading, spacing: 1) {
                                Text(source.name)
                                    .font(.subheadline)
                                    .foregroundStyle(Theme.textPrimary)
                                Text(source.id)
                                    .font(.caption2.monospaced())
                                    .foregroundStyle(.secondary)
                                    .lineLimit(1)
                                if !source.available {
                                    Text("Unavailable — decision retained")
                                        .font(.caption2)
                                        .foregroundStyle(.secondary)
                                }
                            }
                            Spacer(minLength: 0)
                        }
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .listRowInsets(EdgeInsets(
                        top: 5,
                        leading: Theme.Spacing.md,
                        bottom: 5,
                        trailing: Theme.Spacing.md
                    ))
                    .accessibilityIdentifier("access-source-\(source.id)")
                    .accessibilityLabel(source.name)
                    .accessibilityValue(allowed ? "Allowed" : "Blocked")
                    .accessibilityAddTraits(allowed ? .isSelected : [])
                    .disabled(!source.available)
                }
                if shownSources.isEmpty {
                    Text("No sources match that search.")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                }
            }
            // The refusal belongs inside the list that earns it, naming every
            // capability this one list governs.
            if let boundaryError {
                Text(boundaryError)
                    .font(.footnote)
                    .foregroundStyle(Theme.danger)
            }
            // The future-source choice stays inside the list it governs: two
            // lists on one screen would otherwise carry two identical section
            // headers with nothing saying which is whose.
            Text("When you connect a new source")
                .font(.footnote.weight(.semibold))
                .foregroundStyle(Theme.textPrimary)
            futureSourceChoice(allowed: false, title: "Keep it blocked until I allow it")
            futureSourceChoice(allowed: true, title: "Allow it automatically")
        } header: {
            Text(scope.editorTitle)
        }
        .listRowBackground(Theme.bgSecondary)
    }

    private func futureSourceChoice(allowed: Bool, title: String) -> some View {
        Button {
            value.setFutureSourcesAllowed(allowed, knownSourceIds: availableSourceIds)
        } label: {
            HStack(spacing: Theme.Spacing.sm) {
                Image(
                    systemName: value.allowsFutureSources == allowed
                        ? "largecircle.fill.circle"
                        : "circle"
                )
                .foregroundStyle(Theme.accent)
                Text(title)
                    .foregroundStyle(Theme.textPrimary)
                Spacer(minLength: 0)
            }
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(value.allowsFutureSources == allowed ? .isSelected : [])
    }
}

#if DEBUG
@available(iOS 17.0, *)
extension PreviewMocks {
    static let expiredAccessAuthorizationRequest = AccessAuthorizationRequest(
        id: "authorization-expired-preview",
        approvalId: "approval-expired-preview",
        status: "pending",
        clientId: "client-studio",
        clientName: "Studio assistant",
        clientUri: "https://studio-assistant.example.com",
        redirectOrigin: "http://127.0.0.1:49152",
        resource: "https://gateway.example.com/mcp",
        scope: "mcp answer direct",
        expiresAt: 1,
        requiresAnswer: false
    )
}

#Preview("Authorization code") {
    AccessAuthorizationSheet()
        .environment(AppStore())
}

#Preview("Scanned authorization code") {
    AccessAuthorizationSheet(initialCode: "ABCD-EFGH", automaticallyLookup: false)
        .environment(AppStore())
}

#Preview("Permissions · dark · large type") {
    AccessAuthorizationSheet(
        previewRequest: PreviewMocks.accessAuthorizationRequest,
        overview: PreviewMocks.accessOverview,
        mode: .permissions
    )
    .environment(AppStore())
    .preferredColorScheme(.dark)
    .environment(\.dynamicTypeSize, .accessibility2)
}

#Preview("Expired authorization") {
    AccessAuthorizationSheet(
        previewRequest: PreviewMocks.expiredAccessAuthorizationRequest,
        overview: PreviewMocks.accessOverview,
        mode: .review
    )
    .environment(AppStore())
}
#endif

#endif

File: android/app/src/main/kotlin/dev/omnesis/android/ui/access/AccessAuthorizationComponents.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.access

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Checkbox
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.omnesis.android.designsystem.components.OmSpinner
import dev.omnesis.android.designsystem.components.SourceIcon
import dev.omnesis.android.designsystem.components.SourceIconModel
import dev.omnesis.android.designsystem.theme.OmRadius
import dev.omnesis.android.designsystem.theme.OmSpacing
import dev.omnesis.android.designsystem.theme.OmTheme
import dev.omnesis.android.designsystem.theme.OmnesisColors
import dev.omnesis.android.transport.dto.AccessAuthorizationRequest
import dev.omnesis.android.transport.dto.AccessOverview
import dev.omnesis.android.transport.dto.AccessSourceInstance

/**
 * One source list. Checked rows are allowed, a single pair of options decides what happens to a
 * source connected later, and the refusal the list earns is printed inside it, named for every
 * capability the list governs.
 */
@Composable
internal fun SourceRuleEditor(
    title: String,
    scope: AccessSourceScope,
    sources: List<AccessSourceInstance>,
    value: AccessSourceSelectionState,
    iconFor: (String) -> SourceIconModel,
    update: (AccessSourceSelectionState) -> Unit,
) {
    val known = sources.filter { it.available }.mapTo(mutableSetOf()) { it.id }
    var query by remember(title) { mutableStateOf("") }
    val shown = sources.filter { query.isBlank() || it.name.contains(query, true) || it.id.contains(query, true) }
    Column(
        modifier = Modifier.selectableGroup(),
        verticalArrangement = Arrangement.spacedBy(OmSpacing.sm),
    ) {
        SectionTitle(title)
        Text(
            "Checked sources are allowed. You can change this access later.",
            color = OmTheme.colors.textSecondary,
            style = MaterialTheme.typography.bodyMedium,
        )
        if (sources.isEmpty()) {
            Text(
                "No sources are connected.",
                color = OmTheme.colors.textSecondary,
                style = MaterialTheme.typography.bodyMedium,
            )
        } else {
            OutlinedTextField(
                value = query,
                onValueChange = { query = it },
                label = { Text("Search sources") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
            )
            Row(
                Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(OmSpacing.sm),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    "${known.count(value::allows)} of ${known.size} allowed",
                    color = OmTheme.colors.textSecondary,
                    style = MaterialTheme.typography.bodySmall,
                )
                Spacer(Modifier.weight(1f))
                OutlinedButton(onClick = { update(value.allowAll(known)) }) { Text("Allow all") }
                OutlinedButton(onClick = { update(value.blockAll(known)) }) { Text("Block all") }
            }
            shown.forEach { source ->
                SourceSelectionRow(source, value, iconFor) { update(value.setAllowed(source.id, it)) }
            }
            if (shown.isEmpty()) {
                Text(
                    "No sources match that search.",
                    color = OmTheme.colors.textSecondary,
                    style = MaterialTheme.typography.bodyMedium,
                )
            }
        }
        sourceBoundaryError(value, known, scope)?.let { FieldError(it) }
        SubsectionTitle("When you connect a new source")
        RadioRow("Keep it blocked until I allow it", null, !value.futureSourcesAllowed) {
            update(value.setFutureSourcesAllowed(false, known))
        }
        RadioRow("Allow it automatically", null, value.futureSourcesAllowed) {
            update(value.setFutureSourcesAllowed(true, known))
        }
    }
}

@Composable
private fun SourceSelectionRow(
    source: AccessSourceInstance,
    value: AccessSourceSelectionState,
    iconFor: (String) -> SourceIconModel,
    update: (Boolean) -> Unit,
) {
    val enabled = source.available
    val checked = source.available && value.allows(source.id)
    // Many rows of one list being picked from, so each is kept tight: the row
    // is the target, and the name carries it rather than the padding. iOS draws
    // the same row the same way.
    Row(
        Modifier.fillMaxWidth().toggleable(
            value = checked,
            enabled = enabled,
            role = Role.Checkbox,
            onValueChange = update,
        ).padding(vertical = OmSpacing.xs),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Checkbox(checked = checked, onCheckedChange = null, enabled = enabled)
        SourceIcon(iconFor(source.id), size = 20.dp, modifier = Modifier.padding(horizontal = OmSpacing.sm))
        Column(Modifier.weight(1f)) {
            Text(
                source.name,
                fontWeight = FontWeight.Medium,
                style = MaterialTheme.typography.bodyMedium,
            )
            Text(
                source.id,
                color = OmTheme.colors.textSecondary,
                fontFamily = FontFamily.Monospace,
                fontSize = 11.sp,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            if (!source.available) Text("Currently unavailable", color = OmTheme.colors.danger, fontSize = 11.sp)
        }
    }
    HorizontalDivider(color = OmTheme.colors.border)
}

/**
 * One capability, as the same pill everywhere it is named.
 *
 * A granted capability is lit in its own colour and a withheld one is the word gone faint, so a
 * capability's position never moves and a granted capability is never the withheld grey.
 */
@Composable
internal fun CapabilityBadge(
    capability: AccessCapabilitySlot,
    granted: Boolean,
    unreviewed: Boolean = false,
) {
    val tone = capabilityTone(capability, granted, unreviewed)
    val color = tone.color(OmTheme.colors)
    Box(
        Modifier
            .border(
                width = 1.dp,
                color = if (tone == AccessCapabilityTone.WITHHELD) Color.Transparent else color.copy(alpha = .55f),
                shape = RoundedCornerShape(50),
            )
            .padding(horizontal = 7.dp, vertical = 2.dp)
            .semantics {
                contentDescription = capabilityBadgeDescription(capability, granted, unreviewed)
            },
    ) {
        Text(
            capability.label.uppercase(),
            color = if (tone == AccessCapabilityTone.WITHHELD) color.copy(alpha = .55f) else color,
            fontSize = 10.sp,
            fontWeight = FontWeight.SemiBold,
            letterSpacing = 0.4.sp,
            maxLines = 1,
        )
    }
}

/** Three fixed slots in a stable order, so scanning them answers which capabilities are held. */
@Composable
internal fun CapabilityTriad(
    answer: Boolean,
    direct: Boolean,
    notes: Boolean,
    unreviewed: Boolean = false,
) {
    Row(horizontalArrangement = Arrangement.spacedBy(5.dp), verticalAlignment = Alignment.CenterVertically) {
        CapabilityBadge(AccessCapabilitySlot.ANSWER, answer, unreviewed)
        CapabilityBadge(AccessCapabilitySlot.DIRECT, direct)
        CapabilityBadge(AccessCapabilitySlot.NOTES, notes)
    }
}

internal fun AccessCapabilityTone.color(c: OmnesisColors): Color = when (this) {
    AccessCapabilityTone.REVIEWED -> c.success
    AccessCapabilityTone.RAW -> c.danger
    AccessCapabilityTone.WRITE -> c.accent
    AccessCapabilityTone.WARNING -> c.warning
    AccessCapabilityTone.WITHHELD -> c.textMuted
}

/**
 * The last look before access is granted: which connection gets it, which access level it uses,
 * and what that reaches. The permissions are the new level's as the form holds them, the
 * existing level's, or the replaced connection's, stated with the same rows either way. [legacyConnectionName] is the
 * connection a gateway that predates connections reconnects the client to. iOS and the portal
 * state the same rows in the same order.
 */
@Composable
internal fun ReviewStep(
    request: AccessAuthorizationRequest,
    overview: AccessOverview,
    form: AccessAuthorizationForm,
    legacyConnectionName: String? = null,
) {
    val review = remember(form, overview, request, legacyConnectionName) {
        reviewSummary(request, overview, form, legacyConnectionName)
    }
    val permissions = review.permissions
    val answerOn = permissions.answerEnabled || request.requiresAnswer
    val unreviewed = answerOn && permissions.answerRelease == AccessAnswerReleaseChoice.UNREVIEWED
    Column(verticalArrangement = Arrangement.spacedBy(OmSpacing.lg)) {
        SectionTitle("Review access")
        ReviewRow("Connection", review.connectionName)
        review.accessLevel?.let { ReviewRow("Access level", it) }
        review.replaces?.let { ReviewRow("Replaces", it) }
        // The same three badges the capabilities were chosen with, so the thing being
        // approved looks like the thing that was picked.
        CapabilityTriad(
            answer = answerOn,
            direct = permissions.directEnabled,
            notes = permissions.notesEnabled,
            unreviewed = unreviewed,
        )
        permissions.sourceScopes(answerOn).forEach { scope ->
            ReviewRow("${scope.label} sources", sourceSummary(permissions.sources(scope), overview))
        }
        if (answerOn) {
            ReviewRow("Answer privacy", permissions.answerPrivacySummary(overview), warning = unreviewed)
        }
        if (permissions.notesEnabled) {
            ReviewRow("Notes", "Save notes; the agent\u2019s name is recorded")
        }
        sharedLevelFootnote(review.otherConnections)?.let {
            Text(it, color = OmTheme.colors.textSecondary)
        }
        Text(
            "Access can be edited or revoked later from Settings \u2192 Access in the Omnesis Portal.",
            color = OmTheme.colors.textSecondary,
        )
    }
}

@Composable
internal fun WizardActions(
    first: Boolean,
    last: Boolean,
    canContinue: Boolean,
    deciding: Boolean,
    onBack: () -> Unit,
    onContinue: () -> Unit,
    onDeny: () -> Unit,
    onAllow: () -> Unit,
) {
    Row(
        Modifier.fillMaxWidth().padding(horizontal = OmSpacing.lg, vertical = OmSpacing.md),
        horizontalArrangement = Arrangement.spacedBy(OmSpacing.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (first) TextButton(onClick = onDeny, enabled = !deciding) { Text("Deny") }
        else OutlinedButton(onClick = onBack, enabled = !deciding) { Text("Back") }
        Spacer(Modifier.weight(1f))
        Button(
            onClick = if (last) onAllow else onContinue,
            enabled = canContinue && !deciding,
            modifier = Modifier.height(48.dp),
        ) {
            if (deciding) OmSpinner(Modifier.size(18.dp)) else Text(if (last) "Allow access" else "Continue")
        }
    }
}

/**
 * One capability, on or off.
 *
 * [warning] is the capability's own consequence, stated once, here. It is quiet while the
 * capability is one option among three and turns bold and red the moment it applies.
 */
@Composable
internal fun CapabilityCard(
    title: String,
    detail: String,
    checked: Boolean,
    locked: Boolean,
    warning: String? = null,
    onChange: (Boolean) -> Unit,
) {
    val c = OmTheme.colors
    Card(
        modifier = Modifier.fillMaxWidth().toggleable(
            value = checked,
            enabled = !locked,
            role = Role.Checkbox,
            onValueChange = onChange,
        ),
        colors = CardDefaults.cardColors(containerColor = c.bgSecondary),
        border = when {
            !checked -> null
            warning != null -> BorderStroke(1.dp, c.danger)
            else -> BorderStroke(1.dp, c.accent)
        },
        shape = RoundedCornerShape(OmRadius.medium),
    ) {
        Row(Modifier.padding(OmSpacing.lg), verticalAlignment = Alignment.CenterVertically) {
            Checkbox(checked, onCheckedChange = null, enabled = !locked)
            Column(Modifier.padding(start = OmSpacing.sm)) {
                Text(title, fontWeight = FontWeight.SemiBold)
                Text(detail, color = c.textSecondary, style = MaterialTheme.typography.bodyMedium)
                warning?.let {
                    Text(
                        it,
                        color = if (checked) c.danger else c.textSecondary,
                        fontWeight = if (checked) FontWeight.Bold else FontWeight.Normal,
                        style = MaterialTheme.typography.bodyMedium,
                    )
                }
            }
        }
    }
}

/**
 * One option in a group. [warning] is the amber a chosen-but-risky option carries: danger red
 * belongs to Direct alone, and two dangers on one step means neither is read as one.
 */
@Composable
internal fun RadioCard(title: String, detail: String?, selected: Boolean, warning: Boolean = false, onClick: () -> Unit) {
    Card(
        modifier = Modifier.fillMaxWidth().selectable(
            selected = selected,
            role = Role.RadioButton,
            onClick = onClick,
        ),
        colors = CardDefaults.cardColors(containerColor = OmTheme.colors.bgSecondary),
        shape = RoundedCornerShape(OmRadius.medium),
    ) {
        Row(Modifier.padding(OmSpacing.lg), verticalAlignment = Alignment.Top) {
            RadioButton(selected, onClick = null)
            Column(Modifier.padding(start = OmSpacing.sm)) {
                Text(title, fontWeight = FontWeight.SemiBold, color = if (warning) OmTheme.colors.warning else OmTheme.colors.textPrimary)
                detail?.takeIf(String::isNotEmpty)?.let {
                    Text(
                        it,
                        color = if (warning) OmTheme.colors.warning else OmTheme.colors.textSecondary,
                        style = MaterialTheme.typography.bodyMedium,
                    )
                }
            }
        }
    }
}

@Composable
internal fun RadioRow(title: String, detail: String?, selected: Boolean, onClick: () -> Unit) {
    Row(
        Modifier.fillMaxWidth().selectable(selected, role = Role.RadioButton, onClick = onClick)
            .padding(vertical = OmSpacing.xs),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        RadioButton(selected, onClick = null)
        Column(Modifier.padding(start = OmSpacing.sm)) {
            Text(title, fontWeight = FontWeight.Medium)
            detail?.let { Text(it, color = OmTheme.colors.textSecondary, style = MaterialTheme.typography.bodySmall) }
        }
    }
}

@Composable internal fun SubsectionTitle(text: String) =
    Text(text, style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.SemiBold)

@Composable internal fun SectionTitle(text: String) =
    Text(text, style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)

/** A refusal, printed beside the control that earns it rather than at the foot of the step. */
@Composable internal fun FieldError(text: String) =
    Text(text, color = OmTheme.colors.danger, style = MaterialTheme.typography.bodyMedium)

@Composable internal fun WarningCard(text: String) {
    Card(colors = CardDefaults.cardColors(containerColor = OmTheme.colors.danger.copy(alpha = .1f))) {
        Text(text, color = OmTheme.colors.danger, modifier = Modifier.padding(OmSpacing.md))
    }
}

/**
 * One fact of the review, its label above its value. [warning] is the amber an unreviewed
 * release carries here as everywhere else: danger red belongs to Direct alone.
 */
@Composable
private fun ReviewRow(label: String, value: String, warning: Boolean = false) {
    Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(2.dp)) {
        Text(label, color = OmTheme.colors.textSecondary)
        Text(
            value,
            fontWeight = FontWeight.SemiBold,
            color = if (warning) OmTheme.colors.warning else OmTheme.colors.textPrimary,
        )
    }
}

